Isaac/load devotionals

ittullos opened the pull request:
Pretty simple stuff so far in this PR. Would love to know what I'm doing wrong here. Just trying to send an array as a JSON response and getting a weird error.

ittullos committed:
Create /devotionals route

Exposes the list of devotionals to authenticated users.

zbarnard1 commented:
@ittullos What is the error?

EDIT: Dang I should have read the commit message, my bad!

zbarnard1 commented on app/models/devotional.rb:
```diff
@@ -4,4 +4,8 @@ class Devotional < ActiveRecord::Base
   validates :url, presence: true
   validates :img_url, presence: true
+
+  def self.get_devotionals
```
You don't need this, just call Devotional.all in app.rb.

zbarnard1 commented on spec/models/devotional_spec.rb:
```diff
@@ -49,4 +49,12 @@
+    it 'returns a list of devotionals' do
+      create_list(:devotional, 6)
+
+      expect(described_class.get_devotionals.count).to eq(6)
```
Devotional.size would be a little better, that way you don't need to change 6 twice if you ever wanted to update this test.

ittullos replied:
On the right side? (the to eq part?) Says Devotional does not have method "size". I did just change the left side to described_class.count though

zbarnard1 replied:
Ok I'm not crazy, it's a thing.
https://medium.com/unagi/ruby-on-rails-length-vs-size-vs-count-with-examples-a32e17274c79
It might be Rails-specific for some reason but there could be performance implications.

To be honest though, this test is really testing the framework. We are very confident that when you do Devotional.all it will return all of them, so we don't need to test this explicitly here now that we don't have that class method.
The fix here is just to delete the model test. What we want to focus testing on is the API response.

ittullos replied:
Gotcha! 🫡

ittullos commented:
> @ittullos What is the error?

NoMethodError:
       undefined method `attributes' for [#<Devotional id: 1, title: "Daily Bread 1", url: "https://dailybread.com", img_url: "https://dailybread.com/img.jpg">, #<Devotional id: 2, title: "Daily Bread 2", url: "https://dailybread.com", img_url: "https://dailybread.com/img.jpg">, #<Devotional id: 3, title: "Daily Bread 3", url: "https://dailybread.com", img_url: "https://dailybread.com/img.jpg">, #<Devotional id: 4, title: "Daily Bread 4", url: "https://dailybread.com", img_url: "https://dailybread.com/img.jpg">, #<Devotional id: 5, title: "Daily Bread 5", url: "https://dailybread.com", img_url: "https://dailybread.com/img.jpg">, #<Devotional id: 6, title: "Daily Bread 6", url: "https://dailybread.com", img_url: "https://dailybread.com/img.jpg">]:Array

zbarnard1 commented on spec/requests/app_spec.rb:
```diff
@@ -21,4 +21,15 @@
+      get '/devotionals', {}, headers
+
+      expect(last_response.status).to eq(200)
+      expect(parsed_response).to eq(data_object_for(devotionals))
```
Maybe we should change this helper to work for both multiple and single objects instead of having a different one for each, because this won't work.

ittullos replied:
oh the data_object_for? That's where the error is coming from

zbarnard1 replied:
Yep, let's update that to check whether it's getting an array and then it can call `data_objects_for_multiple`. I've hit this error too and it's stupid.

ittullos commented:
looks like it's happening in the data_object_for method

ittullos committed:
Refactor data_object_for method

This commit refactors the data_object_for method in the
api_spec_helpers.rb file. The method now checks the entity to
determine whether it is an array. If it is an array, the method calls
the data_object_for_multiple method. If it is not an array, the method
operates as it did before.

ittullos commented:
Okay, I've refactored the data_object_for method to check for an array and call data_object_for_multiple if so. Let me know what you think. Still a WIP as I still need to write the rake task and test locally.

zbarnard1 commented on spec/support/helpers/api_spec_helpers.rb:
```diff
@@ -10,8 +10,12 @@ def parsed_response
   def data_object_for(entity)
-    entity.attributes.tap do |attributes|
-      convert_timestamps(attributes)
+    if entity.is_a?(Array)
```
Nice!

zbarnard1 commented on spec/models/devotional_spec.rb:
```diff
@@ -49,4 +49,12 @@
     end
   end
+
+  describe '.get_devotionals' do
```
Now that we don't have that class method, do we still need this test?

ittullos replied:
Nope! Good call!

zbarnard1 commented on app/app.rb:
```diff
@@ -34,6 +34,17 @@
+  content_type :json
+  devotionals.to_json
+
+rescue KeyError
```
Under what circumstances would we get a KeyError?

ittullos replied:
Oof so I guess I can't have a KeyError with an array. I'll take a shot at fixing this and see what you think.

zbarnard1 replied:
Let me know if you want to chat through it

zbarnard1 commented on spec/requests/app_spec.rb:
```diff
@@ -21,4 +21,15 @@
+    let!(:devotionals) { create_list(:devotional, 6) }
+
+    it 'returns a list of devotionals' do
+      get '/devotionals', {}, headers
```
Is this route only for authenticated users or is this public? What would happen if an unauthenticated user tried to get here?

ittullos replied:
I would say we only want this route to be accessible to authenticated users since it is behind the login screen. If an unauthenticated user did get here we would want to deny them access.

zbarnard1 replied:
Yep, I agree. Let's have some tests for this case.

ittullos replied:
Okay cool! Will probably need the same for /home right?

zbarnard1 replied:
Yep, but if you wanted to add that in a small followup PR that's cool too. Up to you.

zbarnard1 commented on spec/requests/app_spec.rb:
```diff
@@ -21,4 +21,15 @@
+      get '/devotionals', {}, headers
+
+      expect(last_response.status).to eq(200)
+      expect(parsed_response).to eq(data_object_for(devotionals))
```
Since our `devotionals` here in this test is a list of 6 Devotionals, `data_object_for` should catch any extra or missing elements, so I'd say we're covered as far as that `count` vs `size` thing is concerned. 👍

ittullos commented:
Changed the error handling in the /devotionals route. I'm not really sure if this is the right way to do this, but the test does pass. I'm pretty new to handling errors inside a route, so just following the advice of people online. Let me know what you think

zbarnard1 commented on app/app.rb:
```diff
@@ -34,6 +34,18 @@
+
+    content_type :json
+    devotionals.to_json
+  rescue StandardError => e
```
Cool, this is _ok_... But think about what scenarios would result in an error here, and make sure that we have a test for that scenario.

StandardError is the ancestor/superclass for all kinds of errors (check out
https://www.exceptionalcreatures.com/bestiary.html for some more information on these)

Maybe this rescue block should be applied more generally, rather than just in this block.
Check out how we're handling the KeyError at the bottom of this file. 
Catching errors this way will apply it to ANY request that raises it, which may be an improvement to the API overall!

ittullos replied:
Just changed it back to have a similar structure to /home except using the StandardError. Is that what you meant?

zbarnard1 replied:
No, not quite.
> Maybe this rescue block should be applied more generally, rather than just in this block.
> Check out how we're handling the KeyError at the bottom of this file.

 I was suggesting that instead of only rescuing StandardError within the /devotionals route, we use a strategy more like this, which we have at the bottom of this file: https://github.com/ittullos/pray4tn/blob/5f2cd6e59b180a12f6db16c589ed977bb02db917/app/app.rb#L86

> Catching errors this way will apply it to ANY request that raises it, which may be an improvement to the API overall!

ittullos replied:
Ahh okay gotcha. So I should make a similar structure at the bottom for StandardError? And this happens for the entire file, meaning I wouldn't have to call it here in the method? 

Also is that considered a method at the bottom of the file? Never seen the "error" definition before

zbarnard1 replied:
We can jump on a call and talk through this if you want, it might be easier than commenting back and forth.
This `error SomeErrorClass do` thing is from Sinatra: https://github.com/sinatra/sinatra?tab=readme-ov-file#error-handling
It basically works like a rescue block that applies to all the routes. This way you can avoid duplicating every possible error case in every route block, and you can count on a consistent API response anytime that error occurs.

ittullos committed:
Create Rake task for seeding Devotionals

ittullos commented:
Created the rake task for seeding devotionals. Also changed app.rb to use a similar structure to /home. Let me know if this is what you were looking for or if there's a better way to do it. Thanks!

zbarnard1 commented on spec/tasks/seed_devotionals_spec.rb:
```diff
@@ -0,0 +1,15 @@
+RSpec.describe 'Seeding the database with devotionals', :task do
+  let(:valid_attributes) { attributes_for(:devotional) }
+
+  it 'creates the verses' do
```
We've got a copy-paste error here, this should be 'devotionals' not 'verses'.

ittullos replied:
Good catch!

zbarnard1 commented on Rakefile:
```diff
@@ -20,5 +20,9 @@ namespace :db do
     task :verses => :load_config do
       SeedVerses.run
     end
+    desc "Seed devotionals from /db/seeds/devotionals.yml"
```
⛏️ Let's have a blank line in between the end of the last task and the start of this one to help readability.

ittullos committed:
Refactor error handling on /home route

This gives us a StandarError handling block across the
entire app.

zbarnard1 reviewed (approved):
Great work!
